proposals/1412/client_proxy.md

```diff
@@ -0,0 +1,157 @@
+
+# Proposed Design
+
+## Language Options - Go
+Among many programming language options, I propose using Go/Gin. Go is lightweight, verified in many productions, and has big community. Gin is a web framework for Go and has 64.5K stars in GitHub at the time of writing this. Go programs are compiled into single binary file, which is compact and no additional runtime is needed. There are many other languages with big community such as java, python, or node, but they require additional runtime environment in order to run the software, which is unlike lightweight proxy. Rust/Rocket is another option for lightweight, fast proxy implementation, however, considering the size of the community and the maturity of the project, I would like to propose Go/Gin.
+
+## Design
+![OpenLineage Proxy Sidecar](diagrams/ClientProxySidecar.svg)
+
+### Router(api/v1/lineage)
+This component is a Gin router that creates an API endpoint of `POST /api/v1/lineage`. This will get the actual OpenLineage events payload, validate it via EventValidator, and send it to the EventLogger for local buffering. This is responsible for returning proper response code based on the validation result.
+
+### Event Validator
+This validates that if the event payload is compliant to OpenLineage specification including standard facets. It will pass the invalid events to the FailedEventHandler, which will take care of  putting into the DLQ, logs, or either. In OpenLineage, we will provide a standard interface for it and users could implement their own handler. The implementaion of validator could be as a middleware of Gin router, or a separate component. I will leave the choice to the one who actually implement this.
+
+### Event Logger
+*EventLogger* buffers the OpenLineage events to the temporal local storage. This is responsible for writing events to the log file, and rotating the log based on the configuration. The events must be written to the file by preserving the order of the events that the proxy received. At this point, the event is already validated.
+
+### Event Log Consumer
+This log consumer reads the log written by the *EventLogger*, and send them to the configured backend. It checkpoints the offset of the logs so that it can restart in case of proxy app restart. It is also responsible for retry in case of the backend outage. Backend might not respond to the requests for many possible reasons and this will retry until it succeeds. This will not skip an event and move on to the next in order to avoid data loss.
+
+### Metric Publisher
+*MetricPublisher* gets the data from Event Logger and Event Log Consumer, how many messages are being written to the local storage and how many are being sent to the backend, and it keeps track of the consumer lag metrics. This uses the the *statsd* client so that the metrics can be sent to any *statsd* implementation.
```

Comment:
Do we want to adopt [`OpenMetrics`](https://openmetrics.io/)?

Reply:
That's a good point. The current design does not consider `OpenMetrics` but the `MetricPublisher` will have generic interface for publishing metrics. I think it's worth mentioning possible adoption of `OpenMetrics` in the future.